fix metrics in grpo_fast.py

mnoukhov opened the pull request:
move stop_rate and other metrics before mask_truncated_responses filtering that skews results

add masked_truncated_responses as a metric to log

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Compute stop_rate/good_outputs before truncation filtering, add sequence-length metrics and masked truncation count, and refactor truncation masking into a helper.
> 
> - **Metrics**:
>   - Compute `val/stop_rate` and `good_outputs` before truncation filtering to avoid skew.
>   - Add `val/masked_truncated_responses` and include sequence-length metrics via `calculate_sequence_length_metrics`.
>   - Simplify eval stop rate calculation using boolean mean.
> - **Refactor**:
>   - Extract truncation filtering into `maybe_mask_truncated_completions` helper.
> - **Imports**:
>   - Wire `calculate_sequence_length_metrics` from `rl_utils` and add `GenerationResult` import in `rl_utils`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit f08f7e2b69ef786ca1304762d3ae4f84377f9f64. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

mnoukhov committed:
fix metrics in grpo_fast.py

move stop_rate and other metrics before mask_truncated_responses
filtering that skews results

add masked_truncated_responses as a metric to log

finbarrtimbers reviewed (changes requested):
Can we pull these out into two functions: 

1. calculate_sequence_lengths(result) -> dict[str, float]:
2. maybe_mask_truncated_completions(result) -> tuple[GenerationResult, Batch, dict[str, float]]:
Then it becomes trivial to reorder them.

mnoukhov committed:
separate into own functions @finbarr and fixes for gemini and cursor

cleaner metric implementations
separate out metric implementation into function
separate out whether to truncate into function

mnoukhov committed:
fix type hints

mnoukhov commented on open_instruct/grpo_fast.py:
```diff
@@ -2031,14 +2024,34 @@ def data_preparation_thread(
+        # metrics that need to be calculated before any filtering
+        stop_rate = sum(int(finish_reason == "stop") for finish_reason in result.finish_reasons) / len(
+            result.finish_reasons
+        )
```
done

mnoukhov committed:
Merge branch 'main' of github.com:allenai/open-instruct into michaeln/stop-rate-fix

mnoukhov commented:
/gemini review

saurabh111233212 committed:
Merge branch 'main' into michaeln/stop-rate-fix

mnoukhov committed:
Merge branch 'main' of github.com:allenai/open-instruct into michaeln/stop-rate-fix

mnoukhov committed:
fix buggy name

mnoukhov commented on open_instruct/rl_utils.py:
```diff
@@ -249,3 +250,23 @@ def calculate_advantages_packed(
+    - mean, hist of correct lengths
+    - mean, hist of incorrect results
+    """
+    sequence_lengths = np.array([len(response) for response in result.responses])
```
fixed
